### DISCUSSION TITLE
Make node name a default resource on spans

### DISCUSSION THREAD

**Discussion by fcevado** [open-telemetry/opentelemetry-erlang#402] [Ideas]
For applications with multiple instances it's a pretty useful info to know where a trace was generated.
I'd suggest to set it as a default resource or allowing to pass mfa to resources configuration.

**Comment by bryannaegele** [comment]
Hey @fcevado, you can set these values for the resource via env var which is the easiest way to go about it imo. The env var is `OTEL_RESOURCE_ATTRIBUTES` and you can set any of the values listed in these docs. https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/resource/semantic_conventions

For example:

```
OTEL_RESOURCE_ATTRIBUTES: >
    service.name={{ .Release.Namespace }},service.version={{ .Values.image.tag }},
    container.name={{ .Chart.Name }}, container.image.name={{ .Values.image.repository }},container.image.tag={{ .Values.image.tag }},
    k8s.cluster.name={{ .Values.environment }},k8s.deployment.name={{ .Release.Name }},k8s.namespace.name={{ .Release.Namespace }}
```

**Reply by tsloughter** [reply]
Note you can set any attribute, not just those in the semantic conventions. So if you want to set the Erlang node name as an attribute in the resource you can.

**Reply by fcevado** [reply]
My reasoning was on using the runtime value for that. Sure, I can abstract the value before starting the service and setting it through the envvar option. But i was thinking that `<<"service.instance.id">>` could be set as default to `erlang:node()`.

**Reply by tsloughter** [reply]
Oh, it definitely should. Good catch and thanks for bringing tthis up.

Based on the docs I guess if we detect there is no nodename we should make a uuid https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/resource/semantic_conventions#service

Prefer to not bring in another dependency, though `uuid_erl` would probably never cause conflicts... so may just add support for one version of uuid in the otel lib itself.

**Reply by bryannaegele** [reply]
Can a node run without a name at all? I also don't think the node id could be guaranteed to be globally unique.

**Reply by tsloughter** [reply]
The name would be `nonode@nohost`, I was thinking we would not use that because it would not be unique.

But you are right, it is possible all the nodes of a service use the same name like `release@127.0.0.1`.

Do you know what the default is for like a generated Phoenix release? Not listening for external connections makes the most sense there so I'd assume it is going to be listening on localhost, but don't know for sure.

So I guess instead of doing a check if the name is `nonode@nohost` we should just make it configurable -- obv it is also overridable, but having an option to toggle whether to use the node name or not still seems best.

**Reply by bryannaegele** [reply]
I'm not sure on the Phoenix release. I assume a lot of folks use the same name (k8s DNS), static name, or an IP. It might be difficult to find some default that isn't just a uuid if the name isn't provided

**Reply by bryannaegele** [reply]
Building on your earlier comment, it could definitely be added under a different attribute by default if that's helpful to folks.

**Reply by tsloughter** [reply]
Yea, I'm down with adding a new attribute. I think it'd still be useful to also set `service.instance.id`, but if we are adding a node name attribute I think it would be preferable if we could set `service.instance.id` to the nodename if we can detect it is going to be unique automatically. It won't be perfect but simply not using it if its nonode or `@localhost` may be good enough to make the value useful to most people.

And I mean useful in the sense that APMs will use this attribute and not our node name attribute, so trying our best to make it something meaningful goes a long way I think.

**Reply by fcevado** [reply]
> Do you know what the default is for like a generated Phoenix release?

@tsloughter for releases using `mix release` it gonna be `<appname>@<hostname>` for sure. on k8s `<hostname>` would be the podname. iirc on simple containers network `<hostname>` would be the container name, but it's been a while that i don't work with those. i'll try to run some scenarios here just to make sure.
for releases using [distillery](https://hex.pm/packages/distillery) iirc the rules are the same.

if it's ok i can start working on a PR with a initial approach on that and make adjustments as we figure out the proper way to work with those. what are your thoughts?

**Reply by fcevado** [reply]
> Based on the docs I guess if we detect there is no nodename we should make a uuid

@tsloughter thinking on not adding new dependencies and since uuid in the spec is just an example, wouldn't it be better to use `rand:uniform` just as it's already using for trace id and span id?

**Reply by tsloughter** [reply]
A PR would be great. And yea, I think we can probably get away with just a random number.
